porch/controllers/downstreampackages/api/v1alpha1/downstreampackage_types.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// DownstreamPackage represents an upstream and downstream porch package pair.
+// The upstream package should already exist. The DownstreamPackage controller is
+// responsible for creating the downstream package revisions based on the spec.
+type DownstreamPackage struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DownstreamPackageSpec   `json:"spec,omitempty"`
+	Status DownstreamPackageStatus `json:"status,omitempty"`
+}
+
+func (o *DownstreamPackage) GetSpec() *DownstreamPackageSpec {
+	if o == nil {
+		return nil
+	}
+	return &o.Spec
+}
+
+type AdoptionPolicy string
+type DeletionPolicy string
+
+const (
+	AdoptionPolicyAdoptExisting AdoptionPolicy = "adoptExisting"
+	AdoptionPolicyAdoptNone     AdoptionPolicy = "adoptNone"
+
+	DeletionPolicyDelete = "delete"
+	DeletionPolicyDisown = "disown"
+)
+
+// DownstreamPackageSpec defines the desired state of DownstreamPackage
+type DownstreamPackageSpec struct {
+	Upstream   *Upstream   `json:"upstream,omitempty"`
+	Downstream *Downstream `json:"downstream,omitempty"`
+
+	AdoptionPolicy AdoptionPolicy `json:"adoptionPolicy,omitempty"`
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy,omitempty"`
+}
+
+type Upstream struct {
+	Repo     string `json:"repo,omitempty"`
+	Package  string `json:"package,omitempty"`
```

Comment:
It seems like this essentially identifies a single PackageRevision resource. Should we use the name of the resource instead of `Package` and `Revision`?

Reply:
Given that the resource name of package revisions is auto-generated, that would make this resource very hard to create and read for users, and would require the creation of the package revision prior to the creation of this resource. Generally that's probably going to be the case, but frankly the package revision name is more like a machine generated pointer than something useful. Of course, if this resource is being generated by a controller, this is not really an issue. Usually that is probably the case.

The expected general use of this would be to update the Revision number and trigger upgrades (or downgrades). So keeping these separate makes that easier. This does raise the issue of what happens if someone points this resource at a completely different upstream package, with respect to updates.

Reply:
I see @johnbelamaric's point that if someone is trying to manage a DownstreamPackage resource manually, looking up the package revision resource name for each update can be tedious and perhaps error prone, since it is an auto-generated hash. But the DownstreamPackage is generally going to be created by the UpstreamPolicy controller when it is ready, so this will be less of an issue.

Having the repo/package/revision separate means that the DownstreamPackage controller has put a little more effort into looking up the upstream package revision. If it instead takes the name of the packagerevision resource, the UpstreamPolicy controller will have to do the lookup and pass that to the DownstreamPackage. Either way, we will need to have that lookup somewhere; I think the question is just which controller should we put it in. Personally I don't think that it matters too much, but since the UpstreamPolicy controller is already going to be somewhat complex, maybe it makes sense to delegate the lookup to the DownstreamPackage, and keep the repo/package/revision separate. That said, I don't mind changing it if you feel differently. 

> This does raise the issue of what happens if someone points this resource at a completely different upstream package, with respect to updates.

IIRC porch doesn't let you update a package to a completely different package. We started discussing what that means [here](https://github.com/GoogleContainerTools/kpt/issues/3467#issuecomment-1279221758). I think that in order to completely change the upstream package of the DownstreamPackage, you'd also have to change (or delete) the downstream package target, essentially resulting in a brand new deployment package.